File: java/src/Piece.java
import java.applet.AudioClip;

import javax.swing.ImageIcon;

public abstract class Piece extends Space {

	private boolean selected;
	private boolean color; //false if black, true if white
	private ImageIcon img;
	private AudioClip sound;
	
	public Piece(boolean isWhite, int value, String name, ImageIcon imageIcon, AudioClip sound) {
		super(value, name);
		color = isWhite;
		img = imageIcon;
		this.sound = sound;
	}
	
	public boolean isSelected() {
		return selected;
	}

	public void setSelected(boolean selected) {
		this.selected = selected;
	}
	
	public boolean isWhite() {
		return color;
	}
	
	public ImageIcon getImage() {
		return img;
	}
	
	protected void setImage(ImageIcon imageIcon) {
		this.img = imageIcon;
	}
	
	public void playSound() {
		sound.play();
	}
	
	public String toString() {
		if (color)
			return "w" + super.toString();
		else
			return "b" + super.toString();
	}
}

File: ios/Piece.swift
//
//  Piece.swift
//  Chess
//
//  Created by Jake Present on 6/15/16.
//  Copyright © 2016 Jake Present. All rights reserved.
//

import Foundation
import UIKit

class Piece : Space {
    fileprivate var selected = false
    fileprivate var color:Bool
    fileprivate var img:UIImage!
    
    init(isWhite:Bool, value:Int, type:TypeOfSpace, image:UIImage) {
        
        color = isWhite
        img = image
        
        super.init(value: value, type: type)
    }
    
    required init?(coder aDecoder: NSCoder) {
        
        selected = aDecoder.decodeBool(forKey: "selected")
        color = aDecoder.decodeBool(forKey: "color")
        
        super.init(coder: aDecoder)
        
        let colorString:String
        let typeString:String
        
        if color {
            colorString = "White"
        } else {
            colorString = "Black"
        }
        
        switch self.getType() {
        case .pawn:
            typeString = "Pawn"
        case .bishop:
            typeString = "Bishop"
        case .knight:
            typeString = "Knight"
        case .rook:
            typeString = "Rook"
        case .queen:
            typeString = "Queen"
        case .king:
            typeString = "King"
        case .empty:
            typeString = "Error"
        }
        
        img = UIImage(named: "\(colorString)\(typeString)")
    }
    
    override func encode(with aCoder: NSCoder) {
        super.encode(with: aCoder)
        aCoder.encode(selected, forKey: "selected")
        aCoder.encode(color, forKey: "color")
    }
    
    func isSelected() -> Bool{
        return selected
    }
    
    func setSelected(_ selected:Bool){
        self.selected = selected
    }
    
    func isWhite() -> Bool {
        return color
    }
    
    func getImage() -> UIImage {
        return img
    }
    
    func setImage(_ image:UIImage) {
        img = image
    }
}
